coda-q: tabbed UI

jpwhiting opened the pull request:
* Resolves: # <!-- related github issue -->
* Target version: master 

### Summary


### TODO

### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [x] All commits have Change-Id
- [ ] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

Commit:
CODA-Q: Add a tabbed ui for documents.

When a new document is opened, open in a new tab instead of
a separate window.
Adds a + button which invokes the new document dialog.
When the last tab is closed the application quits.
Show * when a file is modified.
When closing a tab if the file is modified ask user
to save, cancel, or discard.
Also make tabs movable.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: Iff748106caaf81253cfa6a6901e065d11396b716

Commit:
CODA-Q: Make tabs draggable to create new windows.

When a tab is dragged away create a new QMainWindow to hold the tab.
Use a helper function to create the + button on all QTabWidgets.
TODO: Implement merging tabs back into the same window via drag
and drop or menu actions.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: Ide6479f9d26feae521e6a2676fbf8efeee200850

Commit:
CODA-Q: Make removed tabs able to be merged back together.

Dragging a tab from one window can be dropped on another window
and it combines them back together.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: I0594feaffd835c8d15b10546b7442c18fcb9f95d

Commit:
CODA-Q: Use backstage view when + button is clicked.

Instead of showing a functional but ugly "New Document" dialog
when + button is clicked. Open the backstage view.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: I16f3d542f702059957764f4061d72097073e7a2a

Commit:
CODA-Q: Use the current tab for showing the backstageview.

Make each WebView have it's own Window (Maybe these should be merged?)
Connect the + button clicked signal to the Window plusButtonClicked slot
which then tells the bridge of the tab that is currently showing (if
possible) to show it's backstage view.
If there's somehow no visible tab show the functional dialog instead.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: Ie59b110122bcee189a046ac21697f55f93c96ddd

Commit:
CODA-Q: Add an icon to the tabs and window.

Depending on the mime type of the current tab/document
show x-office-(document|spreadsheet|presentation|drawing).svg

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: If47389df617fcdfc4cd56f5b8473d66dcbc12eb5

Commit:
CODA-Q: Prompt each edited file on window close.

Instead of just asking if we should save/discard/cancel prompt
for each modified file if it should be saved, discarded, or cancel
closing.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: I1dc6cbfa05e15ee4515559c2f5863475d67c2f84

Commit:
CODA-Q: Only show the tab bar when there are more than 1 document.

When only 1 tab/document is open in a window hide the tab bar for now.
NOTE: This removes the drop target for merging windows back together,
but that can be fixed separately with a "Merge all windows" menu item
or similar.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: Id46651e41d4239332f7db7fc6a7cb4b03206562b

Commit:
CODA-Q: Add a "Gather Windows" action to the view menu/ribbon.

When Gather Windows is triggered gather all tabs from separate
windows into one primary window.
Adds a windowCount method to only enable the Gather Windows
action if there are multiple tabs.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: I82798bcac466d5a6b4dbcd6e0ea5d1b69a7b7552

stbergmann commented on qt/WebView.cpp:
```diff
@@ -56,8 +65,6 @@
 std::vector<WebView*> WebView::s_instances;
 
-namespace
-{
```
why move the following declarations out of the anonymous namespace?  those declarations appear to be fine with internal linkage?

stbergmann commented on qt/Window.cpp:
```diff
@@ -0,0 +1,159 @@
+    if (closeCallback_) {
+        closeCallback_();
+    }
+
```
One of the commits introduced trailing whitespace here, which a later commit removed again; but which was not ideal when I tried to rebase this commit series onto me local head, as it the first commit caused merge failures, fixing which then had `online`'s git commit hooks kick in and refuse to accept the merge-fixed commit due to the trailing whitespace.  Better force-push the removal of that trailing whitespace into the commit that introduced it.

stbergmann commented on qt/coda-qt.cpp:
```diff
@@ -978,10 +1040,36 @@ QVariant Bridge::cool(const QString& messageStr)
     {
-        // Close the main window associated with this web view
-        if (_webView && _webView->window())
+        // Close the tab if present, otherwise close the main window.
```
(if you introduce terminating punctuation into a comment, my pet-peeve suggestion would be to use a trailing colon instead of a full-stop for comments that precede the commented-on code; that way, it becomes subtly more clear what the comment appertains to)

stbergmann reviewed (commented):
Just a quick first round of comments of what I noticed so far.

When I run `qt/coda-qt`,
* and select "New - Blank Document" it just exits;
* and "Open" and select one existing document, it opens two windows each containing that document;
* and "Open" and multi-select two different documents, they are opened into different windows, not into multiple tabs (and one of them is opened in two windows again)

Commit:
CODA-Q: Fix a few bugs in tabbed ui.

- Properly close the starter tab instead of a whole window.
- Open new documents in new tabs instead of a new window by default.
- Never open or create documents in the starter screen.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: I5a91c768b6e10b7cdf353ae74480277bc616e90f

stbergmann commented on qt/DetachableTabs.cpp:
```diff
@@ -0,0 +1,451 @@
+#include <QTimer>
+#include <QUuid>
+
+// (Temporary debug helper removed)
```
(this looks like a leftover comment that should be removed?)

stbergmann reviewed (commented):
One thing I notice is that now we keep the small size of the initial "Collabora Office - Start" window and open new/existing documents in that small window size, instead of switching to a full-screen document window after showing the initial start window.

(Also, I'm not sure whether we'll want to squash these ten commits when merging, or keep them.  The first nine seem like a nice logical sequence that could be left as such, while the last "CODA-Q: Fix a few bugs in tabbed ui" is a bit of an outlier that might want to be merged into the respective commits it fixes?  I'll leave the decision up to you.)

Quwex commented:
Also Ctrl+W closes all the open tabs instead of the current one.

jpwhiting commented:
> Hi @jpwhiting testing the implementation, on GNOME there seems to be a few pixels width frame around the document at all times. That is more visible when Collabora Desktop is on dark mode. Would it be possible to have the tabs in general to react to the current theme of Collabora Desktop and maybe use the same colors?
> <img alt="image" width="2000" height="1726" src="https://private-user-images.githubusercontent.com/10715461/532291618-a180efef-e4a3-4d92-8a84-81b49c200253.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3MTg4NjcsIm5iZiI6MTc2NzcxODU2NywicGF0aCI6Ii8xMDcxNTQ2MS81MzIyOTE2MTgtYTE4MGVmZWYtZTRhMy00ZDkyLThhODQtODFiNDljMjAwMjUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA2VDE2NTYwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZjZjOTJjNjJhNzVkY2YxZjZkZTA0M2I3YzkxMjM1YWIyYTA5NTc1OWNhMTkxMzgwOWU4YTZhMGEyM2VkMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.y2ESxDxzFHhJ1kytlaU-8yd4fzT7GQRUDSBM5LBut0Y"> <img alt="image" width="2000" height="1726" src="https://private-user-images.githubusercontent.com/10715461/532292160-82108f25-e00c-4801-bd1a-7b9fe71227a0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3MTg4NjcsIm5iZiI6MTc2NzcxODU2NywicGF0aCI6Ii8xMDcxNTQ2MS81MzIyOTIxNjAtODIxMDhmMjUtZTAwYy00ODAxLWJkMWEtN2I5ZmU3MTIyN2EwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA2VDE2NTYwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiYjUwZTQ0NDliYWE0YmM4ODY5NjVhMGVjYzc3MzU4OWY4NjdjYjZiZGQyNjAzYjgyMzlkOWU5MThmYmZjZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0LmipsR3ADlU2e2XZitk1D1Rul0dtRXyW0j4A3CQk-Q">

Thanks for the feedback. That's interesting. I usually run plasma so didn't see that. I've just rebased my branch and changed ctrl-w to close only the current tab. I'll look into stlying the white line/border next and/or just removing that.

jpwhiting commented:
> Also Ctrl+W closes all the open tabs instead of the current one.

Yep, fixed that just now, ctrl-w only closes the current tab now, pushing the changes one moment.

Commit:
CODA-Q: Hide border around tabs.

To prevent a white line appearing around the whole document/window
on GNOME hide the border around the tabwidget completely.
Also use a light theme or dark theme depending on the system theme
for the tabs themselves.

Signed-off-by: Jeremy Whiting <jeremy.whiting@collabora.com>
Change-Id: I112ffeda99082ffc6ba234c1f91faec2d4361b1e

jpwhiting commented:
@Quwex ok, I've removed the line around the window with a Qt stylesheet. It looks fine to me on GNOME now, but I don't use that much. let me know if I missed anything. (I'm not even sure how to choose a different desktop theme there, etc.)

Quwex committed:
fix(app:linux): make ctrl+w close the current tab only

Signed-off-by: Sarper Akdemir <sarper.akdemir@collabora.com>
Change-Id: I540f20a59709c8b92882784ff302417e49a353a3

Darshan-upadhyay1110 commented:
hi @jpwhiting 

Closing this PR since we moved to the Gerrit workflow quite a while ago. For more ref : https://www.collaboraonline.com/blog/open-source-in-action-4/

If you believe this patch is still relevant, can you pretty please port it to [Gerrit](https://gerrit.collaboraoffice.com/). We can review the patch there and proceed accordingly.

Thanks!
